concreate/descriptor.py

```diff
@@ -110,6 +117,164 @@ def _process_labels(self):
                                               'value': self.descriptor['description']})
 
 
+class Resource(object):
+    __SUPPORTED_HASH_ALGORITHMS = ['sha256', 'sha1', 'md5']
+    __check_integrity = True
+
+    @staticmethod
+    def new(resource, base_dir=os.getcwd()):
+        if 'git' in resource:
+            return GitResource(resource)
+        elif 'file' in resource:
+            directory = resource['file']
+            if not os.path.isabs(directory):
+                resource['file'] = os.path.join(base_dir, directory)
+            return FileResource(resource)
+        elif 'url' in resource:
+            return UrlResource(resource)
+        raise ValueError("Resource type is not supported: %s" (resource))
+
+    def __init__(self, descriptor):
+        self.name = descriptor['name']
+        if 'description' in descriptor:
+            self.desription = descriptor['description']
+        self.checksums = {}
+        for algorithm in self.__SUPPORTED_HASH_ALGORITHMS:
+            if algorithm in descriptor:
+                self.checksums[algorithm] = descriptor[algorithm]
+
+    def _copyImpl(self, target):
+        raise NotImplementedError("Implement _copyImpl() for Resource: "
+                                  + self.__module__ + "."
+                                  + type(self).__name__)
+
+    def getTargetFileName(self):
+        return os.path.basename(self.name)
+
+    def copy(self, target_dir=os.getcwd()):
+        target = os.path.join(target_dir, self.getTargetFileName())
+
+        overwrite = False
+        if os.path.exists(target):
+            try:
+                self.__verify(target)
+            except:
+                overwrite = True
+
+        if os.path.exists(target) and not overwrite:
+            logger.debug("Resource '%s' exists, skipping" % self.name)
+            return target
+
+        try:
+            self._copyImpl(target)
+        except Exception as ex:
+            # exception is fatal we be logged before Concreate dies
+            raise ConcreateError('Error copying Resource: %s' % (self.name),
+                                 ex)
+
+        self.__verify(target)
+
+        return target
+
+    def __verify(self, target):
```

Comment:
please do not use two underscore - i think it has different meaning in python. use only one undescore for 'private' stuff.
for more details search for '__double_leading_underscore' in https://www.python.org/dev/peps/pep-0008/

Reply:
I revisited some of this, but those methods I did want to mark as "private" and I believe two leading underscores is the convention, as it prevents subclasses from being able to override them (name mangling).  I also updated some of the other variables/methods to be inline with the style guide (e.g. _ instead of camel case).